第四回課題の提出

Kumagai6824 opened the pull request:
簡単なおみくじプログラムを作成しました。
ご確認お願いいたします。

Kumagai6824 committed:
[Add]Main/FortuneCooker/FortuneLevel/gitignore

yoshi-koyama commented on .gitignore:
```diff
@@ -0,0 +1,8 @@
+*.ipr
+out/
+!**/src/main/**/out/
+!**/src/test/**/out/
```
改行いれておきましょう！

Kumagai6824 replied:
すみません、改行をどこに入れるのか、改行を入れる意味が分かっていません。
参考までにほかの方のも拝見したのですが、あまり情報が得られませんでした。。
「gitignore改行」などで検索してもわからず。もう少し押していただけないでしょうか？

yoshi-koyama commented on src/FortuneLevel.java:
```diff
@@ -0,0 +1,4 @@
+public enum FortuneLevel {
+    大吉, 吉, 凶;
```
Enumは英語で定義しましょう。
```suggestion
    DAIKICHI, KICHI, KYO;
```
または
```suggestion
    EXCELLENT, GOOD, BAD;
```

Kumagai6824 replied:
EXCELLENT（"大吉"）のようにして、書き換えました。

本当はおみくじ結果に"大吉"を表示させたいのですが、うまくいかず。
getterを作成して、dispResultで取得する方法、mainで取得する方法などいろいろ試したのですが、正解にたどり着きませんでした。アドバイスいただけないでしょうか？

yoshi-koyama commented on src/Main.java:
```diff
@@ -0,0 +1,34 @@
+                mapToObj(i -> fortuneList.get(i)).toList();
+
+        System.out.println("<<<WELCOME TO THE FORTUNE COOKER>>>\n");
+        if (result.get(0).getFortuneLevel().equals(FortuneLevel.大吉)) {
```
![badge](https://img.shields.io/badge/review-imo-orange.svg)
毎度`result.get(0)`でListから値を取得するのは手間なので
```java
FortuneCooker yourFortune = result.get(0);
```
としてもいいですね。

Kumagai6824 replied:
修正しました！

TomooAkamatsu commented on src/Main.java:
```diff
@@ -0,0 +1,34 @@
+        Random randomIndexNumber = new Random();
+        List<FortuneCooker> result = randomIndexNumber.
+                ints(1, 0, fortuneList.size()).
+                mapToObj(i -> fortuneList.get(i)).toList();
```
ここの終端処理でtoListとあるのでListが返っていますが、一つだけとわかっているなら他の終端処理でFortuneCookerクラスをそのまま返してもいいかなと思いました。
が、その場合Optionalの処理が必要になるので、理解は難しくなるとは思います...。
もし余力があれば挑戦してみて下さい！
（Optionalは後々必須になるので、今のうちに軽く予習しておくと楽だと思います！）

Kumagai6824 replied:
enumの件を解決させてから、こちら取り組みたいと思います。

Kumagai6824 replied:
`FortuneCooker result = randomIndexNumber.mapToObj(i -> fortuneList.get(i)).findFirst().orElseThrow();`

終端処理を変え、リストではなく、1つの要素になるようにしてみました。ただ、stream自体をしっかり理解できていなく、Optionalも手探りでたまたまできた感じです。optionalで調べるとnullによる例外回避をシンプルに表現する手段として出てくるので、今回の場合どの機能を使うのが良いのか、正直イメージしきれていません。が、上記動くんですが、処理としてはありですか？ご確認お願いします。

TomooAkamatsu replied:
LGTMです！
streamやoptinalは初学者からすると難しい分野ですし、私も頭を悩ませてました。
現時点でここまでふんわりかけていれば十分かと思います。

ざっくり説明すると、
toListだとListで返るので中間処理で一致するものがなくてもListの中身が空という結果で済むのですが、
中身を単体で返す場合は変数がnullになるケースがあり得るので、Optionalによる処理を強制させるようにできています。

Optionalの処理はいくつかありますが、基本的にorElseThrow()でOKだと思います。
nullの場合に例外を投げるメソッドですね。
引数に何も入ってないとNoSuchElementExceptionになります。
こうすることで、NullPointerExceptionを防ぐことができます。(なぜNullPointerを避けるのかは調べてみてください。このあたりは私も含めて未経験からだと実感しにくいところだと思います。)

そういう意味ではNoSuchElementExceptionで終える処理も避けた方が無難ですね
ゆくゆく課題が進んでくると、ここにオリジナル例外などを指定するのですが、現段階ではこちらで十分です！

TomooAkamatsu commented on src/FortuneCooker.java:
```diff
@@ -0,0 +1,29 @@
+public class FortuneCooker {
+    private final Enum FortuneLevel;
+    private final String explanation;
+    private String signOff = "良い一日を。";
```
この辺、適切に書けていて良いですね！

TomooAkamatsu commented on src/FortuneCooker.java:
```diff
@@ -0,0 +1,29 @@
+
+    public void dispResult() {
+        System.out.println
+                ("今日のあなた（PC）の運勢は・・・\n\n      " +
```
ここはテキストブロック使うと楽になるかもしれませんね。

Kumagai6824 replied:
テキストブロックを使って表現してみました。formatted()を使用し、メンバ変数を埋め込む方法を取りましたが、これだと、埋め込む数が多くなった際は、逆に読みにくくなる可能性があるのではないかと思いました。テキストブロックの適切な利用を心がけたいと思います。
今回のテキストブロックの使い方がこれで良さげなのか、、ご確認いただきたいです。

TomooAkamatsu replied:
確認しました！
問題ないと思います。

>逆に読みにくくなる可能性があるのではないかと思いました。

その通りですね。
ぶっちゃけ私自身あまり活用したことがありませんし、今後RESTAPIを作っていくとそもそも体裁を気にして出力することもあまりないと思います笑
こんなものもあるんだな、ぐらいでOKです。

Kumagai6824 committed:
[Modified]
FortuneLevel:Re-define enum eng from jpn
FortuneCooker:Put result.get(0) into yourFortune(variable)
Main:Using Text block for dispResult

Kumagai6824 committed:
[Modified]Main:Changed stream termination
